contracts/bprotocol/scoring/BTokenScore.sol

```diff
@@ -33,48 +73,116 @@ contract BTokenScore is ScoringMachine {
         return keccak256(abi.encodePacked(parent, "coll", cToken));
     }
 
-    function slashedAsset(address cToken) public pure returns (bytes32) {
-        return keccak256(abi.encodePacked(parent, "slashed-debt", cToken));
-    }
-
     // Update Score
     // =============
-    function updateDebtScore(address _user, address cToken, int256 amount) external onlyAvatar {
-        updateScore(user(_user), debtAsset(cToken), amount, now);
+    function updateDebtScore(address _avatar, address cToken, int256 amount) external onlyAvatar {
+        updateScore(user(_avatar), debtAsset(cToken), amount, now);
     }
 
-    function updateCollScore(address _user, address cToken, int256 amount) external onlyAvatar {
-        updateScore(user(_user), collAsset(cToken), amount, now);
+    function updateCollScore(address _avatar, address cToken, int256 amount) external onlyAvatar {
+        updateScore(user(_avatar), collAsset(cToken), amount, now);
     }
 
-    function slashedScore(address _user, address cToken, int256 amount) external {
-        // TODO implement
-        // updateScore(user(_user), slashedAsset(cToken), amount, now);
+    function mul_(uint score, uint multiplier, uint224 index) internal pure returns (uint) {
+        return mul_(mul_(score, multiplier), index);
     }
 
-    // Get Score
-    // ==========
-    function getDebtScore(address _user, address cToken, uint256 time, uint256 spinStart) public view returns (uint) {
-        return getScore(user(_user), debtAsset(cToken), time, spinStart, 0);
+    function updateIndex(address cToken) public {
+        require(endDate > now, "Score: expired");
+        uint224 supplyIndex;
+        uint224 borrowIndex;
+        uint currExchangeRate = ICToken(cToken).exchangeRateCurrent();
+        (supplyIndex,) = comptroller.compSupplyState(cToken);
+        (borrowIndex,) = comptroller.compBorrowState(cToken);
+        snapshot[cToken].exchangeRate = currExchangeRate;
+        snapshot[cToken].supplyIndex = supplyIndex;
+        snapshot[cToken].borrowIndex = borrowIndex;
     }
 
-    function getDebtGlobalScore(address cToken, uint256 time, uint256 spinStart) public view returns (uint) {
-        return getScore(GLOBAL_USER, debtAsset(cToken), time, spinStart, 0);
+    function _getDeltaSupplyIndex(address cToken) internal returns (uint224 deltaSupplyIndex) {
+        uint224 currSupplyIndex;
+        (currSupplyIndex,) = comptroller.compSupplyState(cToken);
+        uint deltaSupplyIndexForCToken = sub_(uint(currSupplyIndex), uint(snapshot[cToken].supplyIndex));
+
+        // NOTICE: supplyIndex takes cToken.totalSupply() which is in cToken quantity
+        // We need the index normalized to underlying token quantity
+        uint deltaExchangeRate = 0;
+        uint deltaSupplyIndexInUint;
+        uint currExchangeRate = ICToken(cToken).exchangeRateCurrent();
+        uint oldExchangeRate = snapshot[cToken].exchangeRate;
+        if(currExchangeRate > oldExchangeRate) {
```

Comment:
the math here is wrong imo. you should always divide the index by current exchange rate (and not delta).
and probably should multiply by 1e18 before diving.